### ISSUE TITLE
Multi match, merge crashing redis

### ISSUE THREAD

**Issue by joshdurbin** [RedisGraph/RedisGraph#1068]
Heya RedisGraph team;

### TLDR

```
MATCH (p:person)-[:transact]->(o:order)-[:contain]->(prd:product)
WITH p, count(prd) as total
MATCH (p)-[:transact]->(o:order)-[:contain]->(prd:product)
WITH p, total, prd, count(o)*1.0 as orders
MERGE (p)-[rated:RATED]->(prd)
ON CREATE SET rated.rating = orders/total
ON MATCH SET rated.rating = orders/total
RETURN p.name, p.name, orders, total, rated.rating
```

is crashing redis w/: `redis-server: /redisgraph/src/execution_plan/ops/op_expand_into.c:160: _handoff: Assertion `Record_GetType(op->r, op->destNodeIdx) == REC_TYPE_NODE' failed.`

### Context

I've written some tooling to generate mock commerce graphs (people, products, orders) that's been iterated on over the last few months. I've had some time lately to focus on the tooling which has helped me identify how / learn about RedisGraph. I've been tinkering with the tooling to get an idea of how I'd integrate things with proper concurrency/behavior etc... and am now at the point where I'm beginning to develop meaningful queries against the data set -- like basic product recommendations.

I've written a bit about this journey [here](https://joshdurbin.net/posts/2020-1-redis-graph-product-recommendation-part-1-data-loading/), but essentially the structure is as follows... There are 3 potential nodes/labels:

- `person`
- `product`
- `order`

... and 4 potential edges/types:

- `addtocart`
- `contain`
- `transact`
- `view`

... where:

- `(:person)-[:view]->(:product)`
- `(:person)-[:addtocart]->(:product)` (records are a subset of `(:person)-[:view]->(:product)`)
- `(:person)-[:transact]->(:order)`
- `(:order)-[:contain]->(:product)` (records are a subset of `(:person)-[:addtocart]->(:product)`)

#### Adapting Neo4j Queries for Product Recommendations

I'm trying to adapt queries in [this gist](https://gist.githubusercontent.com/adam-cowley/79923b538f6851d30e08/raw/abe673493fafba9f2cfd43cba0d7c6b334e6b01e/neo4j-northwind-recommendation) to RedisGraph and the data generated by my mock data generation scripts. My scripts are in the same repo referenced in the aforementioned post, but have changed names (I'll update the post in the next day or so), regardless, they are here:

- [a configurable commerce graph creator](https://github.com/joshdurbin/redis-graph-commerce-poc/blob/master/generateCommerceGraph)
- [a semi-configurable graph query runner](https://github.com/joshdurbin/redis-graph-commerce-poc/blob/master/productRecommendationQueryRunner) (based on the aforementioned gist)

#### Running graph generation scripts

Each of the scripts can be run with the `-h` flag to see their tweakable params and require Java 8 / Groovy 3 to be installed.

Running the `./generateCommerceGraph` script. The script, if you wish to run it generate a sample graph to test the failure below, requires pulling various Maven artifacts at first run, so it can take a few seconds depending on ... 

The [readme](https://github.com/joshdurbin/redis-graph-commerce-poc) in the project is up-to-date and supports tweaking params like the counts, probabilities of edges generated, redis db/host/port, etc...

Run with default params, it will create a graph with counts as follows (give or take):

```
127.0.0.1:6379> graph.query prodrec "match (n) return distinct labels(n), count(n)"
1) 1) "labels(n)"
   2) "count(n)"
2) 1) 1) "order"
      2) (integer) 11686
   2) 1) "person"
      2) (integer) 5000
   3) 1) "product"
      2) (integer) 1000
3) 1) "Query internal execution time: 7.292400 milliseconds"
```

and

```
127.0.0.1:6379> graph.query prodrec "match ()-[e]-() return distinct type(e), count(e)"
1) 1) "type(e)"
   2) "count(e)"
2) 1) 1) "addtocart"
      2) (integer) 63026
   2) 1) "contain"
      2) (integer) 57268
   3) 1) "transact"
      2) (integer) 23372
   4) 1) "view"
      2) (integer) 243034
3) 1) "Query internal execution time: 2909.903400 milliseconds"
(2.91s)
```

##### Crude Approach

The `productRecommendationQueryRunner` script runs the what the author in the gist refers to as a `Content Based Recommendations` on the graph and works rather well. It hasn't been tweaked to remove set purchased products, etc... but it has the basics down.

##### More advanced Recommendations (crashing)

Moving on and to the problem... I'm not trying to run the query for the `Collaborative Filtering` section of the aforementioned gist on my graph and have adapted the query they've supplied:

```
MATCH (c:Customer)-[:PURCHASED]->(o:Order)-[:PRODUCT]->(p:Product)
WITH c, count(p) as total
MATCH (c)-[:PURCHASED]->(o:Order)-[:PRODUCT]->(p:Product)
WITH c, total,p, count(o)*1.0 as orders
MERGE (c)-[rated:RATED]->(p)
ON CREATE SET rated.rating = orders/total
ON MATCH SET rated.rating = orders/total
RETURN c.companyName, p.productName, orders, total, rated.rating
```

...to my graph, which is:

```
MATCH (p:person)-[:transact]->(o:order)-[:contain]->(prd:product)
WITH p, count(prd) as total
MATCH (p)-[:transact]->(o:order)-[:contain]->(prd:product)
WITH p, total, prd, count(o)*1.0 as orders
MERGE (p)-[rated:RATED]->(prd)
ON CREATE SET rated.rating = orders/total
ON MATCH SET rated.rating = orders/total
RETURN p.name, p.name, orders, total, rated.rating
```

...and in doing so I repeatedly crash Redis. The explain for the query is:

```
127.0.0.1:6379> graph.explain prodrec "MATCH (p:person)-[:transact]->(o:order)-[:contain]->(prd:product) WITH p, count(prd) as total MATCH (p)-[:transact]->(o:order)-[:contain]->(prd:product) WITH p, total, prd, count(o)*1.0 as orders MERGE (p)-[rated:RATED]->(prd) ON CREATE SET rated.rating = orders/total ON MATCH SET rated.rating = orders/total RETURN p.name, p.name, orders, total, rated.rating"
 1) "Results"
 2) "    Project"
 3) "        Merge"
 4) "            Aggregate"
 5) "                Expand Into | (o:order)->(prd:product)"
 6) "                    Conditional Traverse | (p)->(o:order)"
 7) "                        Aggregate"
 8) "                            Conditional Traverse | (p:person)->(prd:product)"
 9) "                                Node By Label Scan | (p:person)"
10) "            Expand Into | (p)-[rated:RATED]->(prd)"
11) "                Argument"
12) "            MergeCreate"
13) "                Argument"
```

Thoughts as to what I might be doing wrong or if this is a bug?

Also, probably making all the reference to the mock generation scripts unnecessary, I've included [bulk-operation-result.rdb.tar.gz](https://github.com/RedisGraph/RedisGraph/files/4514182/bulk-operation-result.rdb.tar.gz) of an RDB snapshot after a `save` operation, exported via Redis Insight against the target containerized Redis instance.

Cheers.

#### Crash log

```
redis-server: /redisgraph/src/execution_plan/ops/op_expand_into.c:160: _handoff: Assertion `Record_GetType(op->r, op->destNodeIdx) == REC_TYPE_NODE' failed.


=== REDIS BUG REPORT START: Cut & paste starting from here ===
1:M 22 Apr 2020 06:11:59.850 # Redis 5.0.8 crashed by signal: 11
1:M 22 Apr 2020 06:11:59.850 # Crashed running the instruction at: 0x7f4c0231d611
1:M 22 Apr 2020 06:11:59.850 # Accessing address: (nil)
1:M 22 Apr 2020 06:11:59.850 # Failed assertion: <no assertion failed> (<no file>:0)

------ STACK TRACE ------
EIP:
/lib/x86_64-linux-gnu/libc.so.6(abort+0x1fd)[0x7f4c0231d611]

Backtrace:
redis-server *:6379(logStackTrace+0x32)[0x563cb8a48592]
redis-server *:6379(sigsegvHandler+0x9e)[0x563cb8a48c6e]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12730)[0x7f4c024ce730]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x1fd)[0x7f4c0231d611]
/lib/x86_64-linux-gnu/libc.so.6(+0x2240f)[0x7f4c0231d40f]
/lib/x86_64-linux-gnu/libc.so.6(+0x30102)[0x7f4c0232b102]
/usr/lib/redis/modules/redisgraph.so(+0x1cd0e6)[0x7f4c013080e6]
/usr/lib/redis/modules/redisgraph.so(+0x1c786b)[0x7f4c0130286b]
/usr/lib/redis/modules/redisgraph.so(+0x1c34f9)[0x7f4c012fe4f9]
/usr/lib/redis/modules/redisgraph.so(+0x1c4afc)[0x7f4c012ffafc]
/usr/lib/redis/modules/redisgraph.so(+0x1cbb9c)[0x7f4c01306b9c]
/usr/lib/redis/modules/redisgraph.so(ExecutionPlan_Execute+0x58)[0x7f4c012fc178]
/usr/lib/redis/modules/redisgraph.so(Graph_Query+0x351)[0x7f4c012f6691]
/usr/lib/redis/modules/redisgraph.so(+0x1ebd1d)[0x7f4c01326d1d]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x7fa3)[0x7f4c024c3fa3]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7f4c023f44cf]

------ INFO OUTPUT ------
# Server
redis_version:5.0.8
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:ce75a617c591114f
redis_mode:standalone
os:Linux 4.19.76-linuxkit x86_64
arch_bits:64
multiplexing_api:epoll
atomicvar_api:atomic-builtin
gcc_version:8.3.0
process_id:1
run_id:304844369d91dad6d0e0d3a14692c007239d8519
tcp_port:6379
uptime_in_seconds:409
uptime_in_days:0
hz:10
configured_hz:10
lru_clock:10477615
executable:/data/redis-server
config_file:

# Clients
connected_clients:2
client_recent_max_input_buffer:2
client_recent_max_output_buffer:0
blocked_clients:1

# Memory
used_memory:36811432
used_memory_human:35.11M
used_memory_rss:67522560
used_memory_rss_human:64.39M
used_memory_peak:46109008
used_memory_peak_human:43.97M
used_memory_peak_perc:79.84%
used_memory_overhead:891424
used_memory_startup:824704
used_memory_dataset:35920008
used_memory_dataset_perc:99.81%
allocator_allocated:36197624
allocator_active:39780352
allocator_resident:78159872
total_system_memory:2085785600
total_system_memory_human:1.94G
used_memory_lua:37888
used_memory_lua_human:37.00K
used_memory_scripts:0
used_memory_scripts_human:0B
number_of_cached_scripts:0
maxmemory:0
maxmemory_human:0B
maxmemory_policy:noeviction
allocator_frag_ratio:1.10
allocator_frag_bytes:3582728
allocator_rss_ratio:1.96
allocator_rss_bytes:38379520
rss_overhead_ratio:0.86
rss_overhead_bytes:-10637312
mem_fragmentation_ratio:1.94
mem_fragmentation_bytes:32737632
mem_not_counted_for_evict:0
mem_replication_backlog:0
mem_clients_slaves:0
mem_clients_normal:66616
mem_aof_buffer:0
mem_allocator:jemalloc-5.1.0
active_defrag_running:0
lazyfree_pending_objects:0

# Persistence
loading:0
rdb_changes_since_last_save:30427
rdb_bgsave_in_progress:0
rdb_last_save_time:1587535510
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:-1
rdb_last_cow_size:0
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok
aof_last_cow_size:0

# Stats
total_connections_received:11
total_commands_processed:30449
instantaneous_ops_per_sec:0
total_net_input_bytes:20144895
total_net_output_bytes:3341239
instantaneous_input_kbps:0.00
instantaneous_output_kbps:0.00
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:0
expired_stale_perc:0.00
expired_time_cap_reached_count:0
evicted_keys:0
keyspace_hits:60866
keyspace_misses:1
pubsub_channels:0
pubsub_patterns:0
latest_fork_usec:0
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0

# Replication
role:master
connected_slaves:0
master_replid:1212af7294bb3b60c19e757facfbe6708d293f2b
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:0
second_repl_offset:-1
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:17.670625
used_cpu_user:38.558710
used_cpu_sys_children:0.003566
used_cpu_user_children:0.000000

# Commandstats
cmdstat_graph.QUERY:calls=30436,usec=2300267,usec_per_call=75.58
cmdstat_del:calls=1,usec=94,usec_per_call=94.00
cmdstat_module:calls=5,usec=67,usec_per_call=13.40
cmdstat_type:calls=1,usec=7,usec_per_call=7.00
cmdstat_keys:calls=1,usec=19,usec_per_call=19.00
cmdstat_scan:calls=1,usec=18,usec_per_call=18.00
cmdstat_client:calls=3,usec=31,usec_per_call=10.33
cmdstat_command:calls=1,usec=651,usec_per_call=651.00

# Cluster
cluster_enabled:0

# Keyspace
db0:keys=1,expires=0,avg_ttl=0

------ CLIENT LIST OUTPUT ------
id=60880 addr=172.17.0.1:58474 fd=8 name= age=318 idle=206 flags=N db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=0 obl=0 oll=0 omem=0 events=r cmd=graph.QUERY
id=60893 addr=172.17.0.1:58496 fd=9 name=RedisInsight age=0 idle=0 flags=b db=0 sub=0 psub=0 multi=-1 qbuf=0 qbuf-free=32768 obl=0 oll=0 omem=0 events=r cmd=graph.QUERY

------ REGISTERS ------
1:M 22 Apr 2020 06:11:59.853 # 
RAX:0000000000000000 RBX:0000000000000000
RCX:0000000000000000 RDX:0000000000000000
RDI:0000000000000002 RSI:00007f4bff9350d0
RBP:00007f4c0247fee0 RSP:00007f4bff9351f0
R8 :0000000000000000 R9 :00007f4bff9350d0
R10:0000000000000008 R11:0000000000000246
R12:00007f4c017b3788 R13:00007f4c017b37f8
R14:00000000000000a0 R15:00007f4bd6f01200
RIP:00007f4c0231d611 EFL:0000000000010246
CSGSFS:002b000000000033
1:M 22 Apr 2020 06:11:59.853 # (00007f4bff9351ff) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.853 # (00007f4bff9351fe) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.853 # (00007f4bff9351fd) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.853 # (00007f4bff9351fc) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.853 # (00007f4bff9351fb) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.854 # (00007f4bff9351fa) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.854 # (00007f4bff9351f9) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.854 # (00007f4bff9351f8) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.854 # (00007f4bff9351f7) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.854 # (00007f4bff9351f6) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.854 # (00007f4bff9351f5) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.854 # (00007f4bff9351f4) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.854 # (00007f4bff9351f3) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.854 # (00007f4bff9351f2) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.854 # (00007f4bff9351f1) -> 0000000000000000
1:M 22 Apr 2020 06:11:59.854 # (00007f4bff9351f0) -> 0000000000000020

------ FAST MEMORY TEST ------
1:M 22 Apr 2020 06:11:59.854 # Bio thread for job type #0 terminated
1:M 22 Apr 2020 06:11:59.854 # Bio thread for job type #1 terminated
1:M 22 Apr 2020 06:11:59.854 # Bio thread for job type #2 terminated
*** Preparing to test memory region 563cb8b76000 (2248704 bytes)
*** Preparing to test memory region 563cb9f19000 (135168 bytes)
*** Preparing to test memory region 7f4bd0000000 (135168 bytes)
*** Preparing to test memory region 7f4bd4afe000 (8388608 bytes)
*** Preparing to test memory region 7f4bd52ff000 (8388608 bytes)
*** Preparing to test memory region 7f4bd5b00000 (8388608 bytes)
*** Preparing to test memory region 7f4bd6300000 (101556224 bytes)
*** Preparing to test memory region 7f4be0000000 (3223552 bytes)
*** Preparing to test memory region 7f4be4000000 (4210688 bytes)
*** Preparing to test memory region 7f4be8000000 (3239936 bytes)
*** Preparing to test memory region 7f4bec000000 (3162112 bytes)
*** Preparing to test memory region 7f4bf017e000 (27787264 bytes)
*** Preparing to test memory region 7f4bf1c80000 (22544384 bytes)
*** Preparing to test memory region 7f4bf32ff000 (5242880 bytes)
*** Preparing to test memory region 7f4bf3800000 (8388608 bytes)
*** Preparing to test memory region 7f4bf4000000 (3178496 bytes)
*** Preparing to test memory region 7f4bf8000000 (10485760 bytes)
*** Preparing to test memory region 7f4bf8b7f000 (8388608 bytes)
*** Preparing to test memory region 7f4bf9380000 (8388608 bytes)
*** Preparing to test memory region 7f4bf9b80000 (44564480 bytes)
*** Preparing to test memory region 7f4bfc6b2000 (2621440 bytes)
*** Preparing to test memory region 7f4bfc933000 (8388608 bytes)
*** Preparing to test memory region 7f4bfd134000 (8388608 bytes)
*** Preparing to test memory region 7f4bfd935000 (8388608 bytes)
*** Preparing to test memory region 7f4bfe136000 (8388608 bytes)
*** Preparing to test memory region 7f4bfe937000 (8388608 bytes)
*** Preparing to test memory region 7f4bff138000 (8388608 bytes)
*** Preparing to test memory region 7f4bff939000 (8388608 bytes)
*** Preparing to test memory region 7f4c0013a000 (8388608 bytes)
*** Preparing to test memory region 7f4c0093b000 (8388608 bytes)
*** Preparing to test memory region 7f4c019fd000 (12288 bytes)
*** Preparing to test memory region 7f4c01a00000 (8388608 bytes)
*** Preparing to test memory region 7f4c022f5000 (24576 bytes)
*** Preparing to test memory region 7f4c024b8000 (16384 bytes)
*** Preparing to test memory region 7f4c024d9000 (16384 bytes)
*** Preparing to test memory region 7f4c0266f000 (8192 bytes)
*** Preparing to test memory region 7f4c02672000 (4096 bytes)
*** Preparing to test memory region 7f4c0269c000 (4096 bytes)
.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.%   
```

**Comment by jeffreylovitz** [comment]
Hey @joshdurbin ,

Thanks for the comprehensive writeup! More info is always better.

The crash you're encountering is actually resolved on the RedisGraph master branch (Docker tag `edge`), but we haven't released a new version for `latest` yet.

That said, this does not actually resolve your problem - your query then still returns no results and performs no updates.

The underlying issue is the erroneous introduction of an ExpandInto operation, because the `prd` alias used in the first MATCH clause is thought to still refer to a legitimate entity after the WITH projection in the second MATCH clause.

This is a bug, and we will try to slot some time to solve it soon! In the meantime, you can workaround this by using a different alias:
```
MATCH (p:person)-[:transact]->(:order)-[:contain]->(prd1:product)
WITH p, count(prd1) as total
MATCH (p)-[:transact]->(o:order)-[:contain]->(prd2:product)
WITH p, total, prd2, count(o)*1.0 as orders
MERGE (p)-[rated:RATED]->(prd2)
ON CREATE SET rated.rating = orders/total
ON MATCH SET rated.rating = orders/total
RETURN p.name, p.name, orders, total, rated.rating
```
You can use `edge` or `latest` as you prefer.

On your sample RDB, this creates 28,634 relationships and sets `rated.rating` on each. I cannot tell
if this is the expected output, please let us know!

**Comment by joshdurbin** [comment]
@jeffreylovitz thanks for the quick reply, sorry for the delayed follow-up.

Yeah, that's the expected result for that query. I moved to edge in testing. The follow up queries in that recommendations gist are leveraging currently unsupported OpenCypher functions, so I'm having to adapt some things. Helping me navigate around that error has helped move me forward. Cheers!
